oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,271 @@
+package com.empowerops.volition.ref_oasis
+
+import com.empowerops.volition.dto.*
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.selects.select
+import org.funktionale.either.Either
+import sun.reflect.generics.reflectiveObjects.NotImplementedException
+import java.time.Duration
+
+sealed class SimResult {
+    data class Success(val name: String, val result: Map<String, Double>) : SimResult()
+    data class TimeOut(val name: String) : SimResult()
+    data class TimeOutFailure(val name: String, val exception: String) : SimResult()
+    data class Failure(val name: String, val exception: String) : SimResult()
+}
+
+class OptimizerEndpoint(
+        private val modelService: ModelService
+) : OptimizerGrpc.OptimizerImplBase(), OptimizationControl {
+    var state = State.Idle
+
+    enum class State {
+        Idle,
+        Running
+    }
+
+    override fun changeNodeName(request: NodeNameChangeCommandDTO, responseObserver: StreamObserver<NodeNameChangeResponseDTO>) = responseObserver.consume {
+        val newName = request.newName
+        val newNameSimulation = modelService.simByName[newName]
+        val oldName = request.oldName
+        val target = modelService.simByName[oldName]
+        if (newNameSimulation == null && target != null) {
+            modelService.renameSim(target, newName, oldName)
+            NodeNameChangeResponseDTO.newBuilder().setChanged(true).build()
+        } else {
+            NodeNameChangeResponseDTO.newBuilder().setChanged(false).build()
```

Comment:
Vincent you are a senior software developer at this company. I expect you to remove copy-paste like this before opening a PR. It is not acceptable and you know it. I am not your copy-paste finder.

Reply:
Changed